dali/kernels/signal/resampling_gpu.h

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SIGNAL_RESAMPLING_GPU_H_
+#define DALI_KERNELS_SIGNAL_RESAMPLING_GPU_H_
+
+#include <cuda_runtime.h>
+#include "dali/kernels/signal/resampling.h"
+#include "dali/kernels/signal/resampling_gpu.cuh"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/dynamic_scratchpad.h"
+#include "dali/core/mm/memory.h"
+#include "dali/core/dev_buffer.h"
+#include "dali/core/static_switch.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+
+namespace resampling {
+
+template <typename OutputType = float, typename InputType = OutputType>
+class ResamplerGPU {
+ public:
+  void Initialize(int lobes = 16, int lookup_size = 2048) {
+    windowed_sinc(window_cpu_, lookup_size, lobes);
+    window_gpu_storage_.from_host(window_cpu_.storage);
+    window_gpu_ = window_cpu_;
+    window_gpu_.lookup = window_gpu_storage_.data();
+  }
+
+  KernelRequirements Setup(KernelContext &context, const InListGPU<InputType> &in,
+                           span<const Args> args) {
+    KernelRequirements req;
+    auto out_shape = in.shape;
+    for (int i = 0; i < in.num_samples(); i++) {
+      auto in_sh = in.shape.tensor_shape_span(i);
+      auto out_sh = out_shape.tensor_shape_span(i);
+      auto &arg = args[i];
+      auto out_len = resampled_length(in_sh[0], arg.in_rate, arg.out_rate);
+      auto out_begin = arg.out_begin > 0 ? arg.out_begin : 0;
+      auto out_end = arg.out_end > 0 ? arg.out_end : out_len;
+      if (out_end < out_begin)
+        throw std::invalid_argument(
+            make_string("out_begin can't be larger than out_end. Got out_begin=", out_begin,
+                        ", out_end=", out_end));
+      if (out_end > out_len)
+        throw std::invalid_argument(make_string(
+            "out_end can't be outside of the range of the output signal: [0, ", out_len, ")"));
+      out_sh[0] = out_end - out_begin;
+    }
+    req.output_shapes = {out_shape};
+    return req;
+  }
+
+  void Run(KernelContext &context, const OutListGPU<OutputType> &out,
+           const InListGPU<InputType> &in, span<const Args> args) {
+    if (window_gpu_storage_.empty())
+      Initialize();
```

Comment:
There's a bug - the H2D copy in Initialize happens on stream 0 and may not complete before the processing starts here.